### PR TITLE
Integrated validation into graphalytics-core

### PR DESCRIPTION
**Pull request by stijnh**
Added ability for core to perform automatic validation of output after each benchmark. The validation is performed according to the rules specified in the Graphalytics specifications (i.e., correct handling of floating-point precision, equivalence, etc.). Validation can be enabled using the flag "benchmark.run.validation-requried" and the reference output should be placed in directory "benchmark.run.validation-directory" in files named "[dataset name]-[algorithm name]" (e.g., wiki-Talk-PR).

### CODE REVIEW

File: graphalytics-core/src/main/java/nl/tudelft/graphalytics/validation/rule/EpsilonValidationRule.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+	@Override
+	public boolean match(Double outputValue, Double correctValue) {
+		try {
+			double a = outputValue;
+			double b = correctValue;
+
+			if (Double.isNaN(a) && Double.isNaN(b)) {
+				return true;
+			} else if (Double.isInfinite(a) && Double.isInfinite(b) && a * b < 0) {
```
Comment (thegeman) [inline graphalytics-core/src/main/java/nl/tudelft/graphalytics/validation/rule/EpsilonValidationRule.java]: Are two infinite values not comparable using `==`?

I believe the multiplication should yield a positive number if the signs are equal (`a * b > 0`). Better yet, check the signs more explicitly (I assume this works for pos/neg infinity): `Math.signum(a) == Math.signum(b)`.

File: graphalytics-core/src/main/java/nl/tudelft/graphalytics/validation/rule/EpsilonValidationRule.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+				return true;
+			} else if (Math.abs(a - b) < COMPARISON_THRESHOLD * a) {
```
Comment (thegeman) [inline graphalytics-core/src/main/java/nl/tudelft/graphalytics/validation/rule/EpsilonValidationRule.java]: Minor detail: in the specs we state that `Math.abs(output - correct) < EPSILON * correct`, so the last `a` should become a `b`. Not likely to ever matter, but I would rather be consistent.

File: graphalytics-core/src/main/java/nl/tudelft/graphalytics/validation/rule/EpsilonValidationRule.java (hunk 1 above)
Comment (stijnh) [inline graphalytics-core/src/main/java/nl/tudelft/graphalytics/validation/rule/EpsilonValidationRule.java]: You are correct, infinite values can indeed be compared using ==, only NaN values cannot apparently. Updated branch.

File: graphalytics-core/src/main/java/nl/tudelft/graphalytics/validation/rule/EpsilonValidationRule.java (hunk 2 above)
Comment (stijnh) [inline graphalytics-core/src/main/java/nl/tudelft/graphalytics/validation/rule/EpsilonValidationRule.java]: You are correct. Updated branch.

**Comment by thegeman** [comment]
ok to test

**Pull request merged**